File: 2024/day09/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;

# Advent of Code 2024 Day 9: Disk Fragmenter
#
# Compact a fragmented disk by moving file blocks to fill gaps.
# Part 1: Move blocks one at a time from end to leftmost free space
# Part 2: Move whole files to leftmost span that fits

sub parse_disk_map {
    my ($filename) = @_;

    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    my $disk_map = <$fh>;
    close $fh;
    chomp $disk_map;

    my @blocks;
    my $file_id = 0;
    my $is_file = 1;

    for my $digit (split //, $disk_map) {
        my $length = int($digit);
        if ($is_file) {
            push @blocks, ($file_id) x $length;
            $file_id++;
        } else {
            push @blocks, (-1) x $length;  # -1 represents free space
        }
        $is_file = !$is_file;
    }

    return \@blocks;
}

sub compact_blocks {
    my ($blocks_ref) = @_;
    my @blocks = @$blocks_ref;  # Copy the array

    my $left = 0;
    my $right = $#blocks;

    while ($left < $right) {
        # Find leftmost free space
        while ($left < $right && $blocks[$left] != -1) {
            $left++;
        }
        # Find rightmost file block
        while ($left < $right && $blocks[$right] == -1) {
            $right--;
        }

        if ($left < $right) {
            # Swap
            $blocks[$left] = $blocks[$right];
            $blocks[$right] = -1;
            $left++;
            $right--;
        }
    }

    return \@blocks;
}

sub calculate_checksum {
    my ($blocks_ref) = @_;
    my @blocks = @$blocks_ref;

    my $checksum = 0;
    for my $pos (0 .. $#blocks) {
        if ($blocks[$pos] != -1) {
            $checksum += $pos * $blocks[$pos];
        }
    }

    return $checksum;
}

sub part1 {
    my $blocks = parse_disk_map('../input.txt');
    my $compacted = compact_blocks($blocks);
    return calculate_checksum($compacted);
}

sub part2 {
    my $blocks_ref = parse_disk_map('../input.txt');
    my @blocks = @$blocks_ref;

    # Find all files: file_id -> [start_pos, length]
    my %files;
    my $i = 0;
    while ($i < @blocks) {
        if ($blocks[$i] != -1) {
            my $file_id = $blocks[$i];
            my $start = $i;
            while ($i < @blocks && $blocks[$i] == $file_id) {
                $i++;
            }
            $files{$file_id} = [$start, $i - $start];
        } else {
            $i++;
        }
    }

    # Find max file ID
    my $max_file_id = (sort { $b <=> $a } keys %files)[0];

    # Process files in decreasing order of file ID
    for (my $file_id = $max_file_id; $file_id >= 0; $file_id--) {
        next unless exists $files{$file_id};

        my ($start, $length) = @{$files{$file_id}};

        # Find leftmost span of free space that fits this file
        # Must be to the left of current position
        my $free_start = undef;
        $i = 0;
        while ($i < $start) {
            if ($blocks[$i] == -1) {
                # Count consecutive free blocks
                my $span_start = $i;
                my $span_length = 0;
                while ($i < $start && $blocks[$i] == -1) {
                    $span_length++;
                    $i++;
                }
                if ($span_length >= $length) {
                    $free_start = $span_start;
                    last;
                }
            } else {
                $i++;
            }
        }

        # Move file if we found a suitable span
        if (defined $free_start) {
            # Clear old position
            for my $j ($start .. $start + $length - 1) {
                $blocks[$j] = -1;
            }
            # Write to new position
            for my $j ($free_start .. $free_start + $length - 1) {
                $blocks[$j] = $file_id;
            }
            # Update file position
            $files{$file_id} = [$free_start, $length];
        }
    }

    return calculate_checksum(\@blocks);
}

say "Part 1: ", part1();
say "Part 2: ", part2();

File: 2024/day09/zig/solution.zig
const std = @import("std");

const FileInfo = struct {
    start: usize,
    length: usize,
};

fn parseDiskMap(allocator: std.mem.Allocator, filename: []const u8) !std.ArrayList(i32) {
    const file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    // Trim newline/whitespace
    const disk_map = std.mem.trim(u8, content, &std.ascii.whitespace);

    var blocks: std.ArrayList(i32) = .{};
    var file_id: i32 = 0;
    var is_file = true;

    for (disk_map) |ch| {
        const length = ch - '0';
        var i: u8 = 0;
        while (i < length) : (i += 1) {
            if (is_file) {
                try blocks.append(allocator, file_id);
            } else {
                try blocks.append(allocator, -1); // -1 represents free space
            }
        }
        if (is_file) {
            file_id += 1;
        }
        is_file = !is_file;
    }

    return blocks;
}

fn compactBlocks(allocator: std.mem.Allocator, blocks: []const i32) !std.ArrayList(i32) {
    var result: std.ArrayList(i32) = .{};
    try result.appendSlice(allocator, blocks);

    var left: usize = 0;
    var right: usize = result.items.len - 1;

    while (left < right) {
        // Find leftmost free space
        while (left < right and result.items[left] != -1) {
            left += 1;
        }
        // Find rightmost file block
        while (left < right and result.items[right] == -1) {
            right -= 1;
        }

        if (left < right) {
            // Swap
            result.items[left] = result.items[right];
            result.items[right] = -1;
            left += 1;
            right -= 1;
        }
    }

    return result;
}

fn calculateChecksum(blocks: []const i32) u64 {
    var checksum: u64 = 0;
    for (blocks, 0..) |file_id, pos| {
        if (file_id != -1) {
            checksum += @as(u64, pos) * @as(u64, @intCast(file_id));
        }
    }
    return checksum;
}

fn part1(allocator: std.mem.Allocator) !u64 {
    var blocks = try parseDiskMap(allocator, "../input.txt");
    defer blocks.deinit(allocator);

    var compacted = try compactBlocks(allocator, blocks.items);
    defer compacted.deinit(allocator);

    return calculateChecksum(compacted.items);
}

fn part2(allocator: std.mem.Allocator) !u64 {
    var blocks = try parseDiskMap(allocator, "../input.txt");
    defer blocks.deinit(allocator);

    // Find all files: file_id -> (start_pos, length)
    var files = std.AutoHashMap(i32, FileInfo).init(allocator);
    defer files.deinit();

    var i: usize = 0;
    while (i < blocks.items.len) {
        if (blocks.items[i] != -1) {
            const file_id = blocks.items[i];
            const start = i;
            while (i < blocks.items.len and blocks.items[i] == file_id) {
                i += 1;
            }
            try files.put(file_id, FileInfo{ .start = start, .length = i - start });
        } else {
            i += 1;
        }
    }

    // Find max file ID
    var max_file_id: i32 = 0;
    var it = files.keyIterator();
    while (it.next()) |key| {
        if (key.* > max_file_id) {
            max_file_id = key.*;
        }
    }

    // Process files in decreasing order of file ID
    var file_id = max_file_id;
    while (file_id >= 0) : (file_id -= 1) {
        const file_info = files.get(file_id) orelse continue;
        const start = file_info.start;
        const length = file_info.length;

        // Find leftmost span of free space that fits this file
        // Must be to the left of current position
        var free_start: ?usize = null;
        i = 0;
        while (i < start) {
            if (blocks.items[i] == -1) {
                // Count consecutive free blocks
                const span_start = i;
                var span_length: usize = 0;
                while (i < start and blocks.items[i] == -1) {
                    span_length += 1;
                    i += 1;
                }
                if (span_length >= length) {
                    free_start = span_start;
                    break;
                }
            } else {
                i += 1;
            }
        }

        // Move file if we found a suitable span
        if (free_start) |fs| {
            // Clear old position
            var j: usize = start;
            while (j < start + length) : (j += 1) {
                blocks.items[j] = -1;
            }
            // Write to new position
            j = fs;
            while (j < fs + length) : (j += 1) {
                blocks.items[j] = file_id;
            }
            // Update file position in map
            try files.put(file_id, FileInfo{ .start = fs, .length = length });
        }
    }

    return calculateChecksum(blocks.items);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const p1 = try part1(allocator);
    std.debug.print("Part 1: {d}\n", .{p1});

    const p2 = try part2(allocator);
    std.debug.print("Part 2: {d}\n", .{p2});
}
